src/base.c

```diff
@@ -295,10 +295,14 @@ base_block_alloc(tsdn_t *tsdn, base_t *base, ehooks_t *ehooks, unsigned ind,
 	return block;
 }
 
-static ehooks_t *
-base_ehooks_get_for_metadata(base_t *base) {
-	return base->metadata_use_hooks ? &base->ehooks :
-	    (ehooks_t *)&ehooks_default_extent_hooks;
```

Comment:
Good catch! This is indeed an invalid conversion.

I think we can make the fix simpler though; the current PR attempts to workaround the difference between ehooks_t and extent_hooks_t by adding an ehooks_t on the stack. However we generally avoid this pattern, unless its forced by bootstrapping (e.g. the `fake_ehooks` in base_new).

Instead we can simply add another ehooks_t in base_s, `ehooks_base` to replace `bool metadata_use_hooks`. Then the added ehooks_base can be initialized to either the same as `ehooks`, or pointing to the default hooks, based on the `metadata_use_hooks` pass into `base_new`.

The other request is, please squash everything into a single commit, including the test cases. Then you can `git push -f` to update the PR.